File: app/src/main/java/de/robinrehbein/punkt/game/TimingGame.kt
package de.robinrehbein.punkt.game

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sin
import kotlin.random.Random

/**
 * Pure-Kotlin Engine für das "Stopp"-Spielprinzip: Timing-Präzision.
 *
 * Der Punkt läuft automatisch auf einer Kreisbahn. Irgendwo auf der Bahn
 * liegt eine Zielzone — ein Tap, während der Punkt in der Zone ist,
 * zählt einen Treffer: Die Laufrichtung dreht um, die Zone springt an
 * eine neue Position, das Tempo steigt und die Zone schrumpft. Ein Tap
 * außerhalb der Zone oder ein Überfahren der Zone ohne Tap ist sofort
 * das Ende. Perfekte Treffer (Zonenmitte) zählen doppelt.
 *
 * Damit es nicht langweilig wird, schalten sich mit steigendem Score
 * "Twists" frei, die pro Zone zufällig gemischt werden:
 * - PULSE (ab 5): Die Zone pulsiert, wächst und schrumpft rhythmisch.
 * - DRIFT (ab 10): Die Zone wandert langsam auf der Bahn.
 * - GHOST (ab 15): Der Punkt blinkt periodisch weg.
 * - FAKE (ab 20): Eine Köder-Zone auf dem Weg — Tap darin ist tödlich.
 * - CHAIN (ab 25): Zwei Zonen direkt nacheinander ohne Richtungswechsel.
 *
 * Winkel sind in Radiant, Geschwindigkeiten in Radiant pro Sekunde.
 */
class TimingGame(private val random: Random = Random.Default) {

    enum class Phase { READY, RUNNING, DYING, OVER }

    enum class Twist { PULSE, DRIFT, GHOST, FAKE, CHAIN }

    sealed interface GameEvent {
        data object Started : GameEvent
        data object Hit : GameEvent
        data object PerfectHit : GameEvent
        data object ChainNext : GameEvent
        data class TwistUnlocked(val twist: Twist) : GameEvent
        data object Died : GameEvent
        data object Settled : GameEvent
    }

    var phase: Phase = Phase.READY
        private set

    /** Position des Punkts auf der Bahn. */
    var angle: Float = 0f
        private set

    /** Laufrichtung: +1 = im Uhrzeigersinn, -1 = dagegen. */
    var direction: Int = 1
        private set

    var zoneCenter: Float = 1.8f
        private set

    /** Basisbreite der Zone (halb); die effektive Breite kann pulsieren. */
    var zoneHalfWidth: Float = BASE_ZONE_HALF
        private set

    var score: Int = 0
        private set

    var elapsed: Float = 0f
        private set

    /** Zeit seit dem letzten Treffer, für Animationen. */
    var timeSinceHit: Float = 99f
        private set

    /** War der letzte Treffer ein perfekter? Für Anzeige-Effekte. */
    var lastHitPerfect: Boolean = false
        private set

    /** Die für die aktuelle Zone aktiven Twists. */
    val activeTwists = mutableSetOf<Twist>()

    /** Köder-Zone (nur relevant, wenn FAKE aktiv ist). */
    var hasFakeZone: Boolean = false
        private set
    var fakeZoneCenter: Float = 0f
        private set

    /** Wie viele Ketten-Zonen nach der aktuellen noch folgen. */
    var chainRemaining: Int = 0
        private set

    /** Drift-Richtung relativ zur Laufrichtung: +1 = flieht, -1 = kommt entgegen. */
    private var driftSign: Int = 1

    /** Bereits angekündigte Twists (Banner nur einmal pro Lauf). */
    private val announcedTwists = mutableSetOf<Twist>()

    /**
     * Nur für Tests/Debug: erzwingt ein festes Twist-Set für alle
     * folgenden Zonen statt der zufälligen Auswahl.
     */
    var twistOverride: Set<Twist>? = null

    /**
     * Events aus tap() werden gepuffert und mit dem nächsten update()
     * ausgeliefert, damit der UI-Loop eine einzige Event-Quelle hat.
     */
    private val pendingEvents = mutableListOf<GameEvent>()

    /** Relative Position des Punkts zur Zone: negativ = davor, 0 = Mitte. */
    fun relativeToZone(): Float = wrapToPi(direction * (angle - zoneCenter))

    /** Effektive halbe Zonenbreite — pulsiert, wenn PULSE aktiv ist. */
    fun effectiveZoneHalf(): Float {
        if (Twist.PULSE !in activeTwists) return zoneHalfWidth
        val pulse = PULSE_MIN_SHARE + (1f - PULSE_MIN_SHARE) *
            (0.5f + 0.5f * sin(elapsed * PULSE_SPEED))
        return zoneHalfWidth * pulse
    }

    /** Steht der Punkt gerade in der (effektiven) Zielzone? */
    val isInZone: Boolean get() = abs(relativeToZone()) <= effectiveZoneHalf()

    /** Ist der Punkt gerade sichtbar? Blinkt nur im GHOST-Twist. */
    val isDotVisible: Boolean
        get() = phase != Phase.RUNNING || Twist.GHOST !in activeTwists ||
            (elapsed * GHOST_BLINK_SPEED) % 1f < GHOST_VISIBLE_SHARE

    fun currentSpeed(): Float =
        (BASE_SPEED + score * SPEED_PER_HIT).coerceAtMost(MAX_SPEED)

    /**
     * Verarbeitet einen Tap. In READY startet er den Lauf, in RUNNING ist
     * er der Stopp-Versuch, in OVER (nach kurzer Sperre gegen Wut-Taps)
     * geht es zurück in den READY-Zustand.
     */
    fun tap(): GameEvent? {
        val event: GameEvent? = when (phase) {
            Phase.READY -> {
                phase = Phase.RUNNING
                elapsed = 0f
                spawnZone()
                GameEvent.Started
            }
            Phase.RUNNING -> {
                val rel = relativeToZone()
                val half = effectiveZoneHalf()
                if (abs(rel) <= half) {
                    val perfect = abs(rel) <= half * PERFECT_SHARE
                    registerHit(perfect)
                    if (perfect) GameEvent.PerfectHit else GameEvent.Hit
                } else {
                    // Auch ein Tap in der Fallen-Zone landet hier: Sie ist
                    // mechanisch einfach "daneben" — ihre Gefahr ist optisch.
                    die()
                    GameEvent.Died
                }
            }
            Phase.DYING -> null
            Phase.OVER -> {
                if (elapsed >= RESTART_LOCK_SECONDS) {
                    reset()
                }
                null
            }
        }
        if (event != null) pendingEvents.add(event)
        return event
    }

    /** Setzt alles auf den READY-Zustand zurück (Rekord bleibt beim Store). */
    fun reset() {
        phase = Phase.READY
        angle = 0f
        direction = 1
        zoneCenter = 1.8f
        zoneHalfWidth = BASE_ZONE_HALF
        score = 0
        elapsed = 0f
        timeSinceHit = 99f
        lastHitPerfect = false
        activeTwists.clear()
        hasFakeZone = false
        chainRemaining = 0
        announcedTwists.clear()
        pendingEvents.clear()
    }

    /**
     * Schreibt einen Frame fort und liefert die dabei aufgetretenen Events.
     */
    fun update(deltaSeconds: Float): List<GameEvent> {
        val dt = deltaSeconds.coerceIn(0f, MAX_DELTA)
        elapsed += dt
        timeSinceHit += dt
        val events = mutableListOf<GameEvent>()
        events.addAll(pendingEvents)
        pendingEvents.clear()

        when (phase) {
            Phase.READY -> {
                angle = wrapTwoPi(angle + direction * READY_SPEED * dt)
            }
            Phase.RUNNING -> {
                angle = wrapTwoPi(angle + direction * currentSpeed() * dt)
                if (Twist.DRIFT in activeTwists) {
                    zoneCenter = wrapTwoPi(
                        zoneCenter + direction * driftSign * DRIFT_SPEED * dt
                    )
                }
                // Zone ohne Tap überfahren → vorbei. Geprüft wird gegen die
                // volle Basisbreite, damit eine pulsierende Zone fair bleibt.
                if (relativeToZone() > zoneHalfWidth + PASS_BUFFER) {
                    die()
                    events.add(GameEvent.Died)
                }
            }
            Phase.DYING -> {
                // Kurzer Freeze für Flash und Shake, dann steht das Ergebnis.
                if (elapsed >= DEATH_FREEZE_SECONDS) {
                    phase = Phase.OVER
                    elapsed = 0f
                    events.add(GameEvent.Settled)
                }
            }
            Phase.OVER -> Unit
        }
        return events
    }

    private fun registerHit(perfect: Boolean) {
        score += if (perfect) PERFECT_SCORE else 1
        timeSinceHit = 0f
        lastHitPerfect = perfect
        zoneHalfWidth = (BASE_ZONE_HALF - score * ZONE_SHRINK_PER_HIT)
            .coerceAtLeast(MIN_ZONE_HALF)

        if (chainRemaining > 0) {
            // Ketten-Zone: gleiche Richtung, die nächste kommt sofort.
            chainRemaining--
            hasFakeZone = false
            activeTwists.remove(Twist.FAKE)
            spawnChainZone()
            pendingEvents.add(GameEvent.ChainNext)
        } else {
            direction = -direction
            spawnZone()
        }
    }

    private fun spawnZone() {
        val distance = MIN_ZONE_DISTANCE +
            random.nextFloat() * (MAX_ZONE_DISTANCE - MIN_ZONE_DISTANCE)
        zoneCenter = wrapTwoPi(angle + direction * distance)
        chooseTwists()

        driftSign = if (random.nextBoolean()) 1 else -1
        chainRemaining = if (Twist.CHAIN in activeTwists) CHAIN_LENGTH else 0

        hasFakeZone = false
        if (Twist.FAKE in activeTwists) {
            val maxFakeDistance = distance - zoneHalfWidth * 3f
            if (maxFakeDistance > FAKE_MIN_DISTANCE) {
                val fakeDistance = FAKE_MIN_DISTANCE +
                    random.nextFloat() * (maxFakeDistance - FAKE_MIN_DISTANCE)
                fakeZoneCenter = wrapTwoPi(angle + direction * fakeDistance)
                hasFakeZone = true
            }
        }
    }

    /** Folge-Zone einer Kette: näher dran, keine neue Twist-Auswahl. */
    private fun spawnChainZone() {
        val distance = CHAIN_MIN_DISTANCE +
            random.nextFloat() * (CHAIN_MAX_DISTANCE - CHAIN_MIN_DISTANCE)
        zoneCenter = wrapTwoPi(angle + direction * distance)
    }

    private fun chooseTwists() {
        activeTwists.clear()
        val override = twistOverride
        if (override != null) {
            activeTwists.addAll(override)
            return
        }

        val unlocked = Twist.entries.filter { score >= unlockScore(it) }

        // Ein frisch freigeschalteter Twist wird garantiert gezeigt
        // und einmalig angekündigt.
        val fresh = unlocked.firstOrNull { it !in announcedTwists }
        if (fresh != null) {
            activeTwists.add(fresh)
            announcedTwists.add(fresh)
            pendingEvents.add(GameEvent.TwistUnlocked(fresh))
        }

        val shuffled = unlocked.shuffled(random)
        for (twist in shuffled) {
            if (activeTwists.size >= MAX_ACTIVE_TWISTS) break
            if (twist in activeTwists) continue
            if (random.nextFloat() < TWIST_PROBABILITY) {
                activeTwists.add(twist)
            }
        }
    }

    private fun die() {
        if (phase != Phase.RUNNING) return
        phase = Phase.DYING
        elapsed = 0f
    }

    companion object {
        const val MAX_DELTA = 1f / 30f

        // Tempo (Radiant pro Sekunde)
        const val BASE_SPEED = 2.4f
        const val SPEED_PER_HIT = 0.07f
        const val MAX_SPEED = 5.2f
        const val READY_SPEED = 1.2f

        // Zielzone (Radiant)
        const val BASE_ZONE_HALF = 0.40f
        const val ZONE_SHRINK_PER_HIT = 0.006f
        const val MIN_ZONE_HALF = 0.15f
        const val PERFECT_SHARE = 0.35f
        const val PASS_BUFFER = 0.05f
        const val MIN_ZONE_DISTANCE = 1.1f
        const val MAX_ZONE_DISTANCE = 2.8f

        // Scoring
        const val PERFECT_SCORE = 2

        // Twists
        const val MAX_ACTIVE_TWISTS = 2
        const val TWIST_PROBABILITY = 0.45f
        const val PULSE_SPEED = 5f
        const val PULSE_MIN_SHARE = 0.62f
        const val DRIFT_SPEED = 0.35f
        const val GHOST_BLINK_SPEED = 1.6f
        const val GHOST_VISIBLE_SHARE = 0.62f
        const val FAKE_MIN_DISTANCE = 0.55f
        const val CHAIN_LENGTH = 1
        const val CHAIN_MIN_DISTANCE = 1.0f
        const val CHAIN_MAX_DISTANCE = 1.8f

        const val DEATH_FREEZE_SECONDS = 0.5f
        const val RESTART_LOCK_SECONDS = 0.55f

        /** Ab welchem Score ein Twist ins Spiel kommt. */
        fun unlockScore(twist: Twist): Int = when (twist) {
            Twist.PULSE -> 5
            Twist.DRIFT -> 10
            Twist.GHOST -> 15
            Twist.FAKE -> 20
            Twist.CHAIN -> 25
        }

        private const val TWO_PI = (2 * PI).toFloat()

        /** Normalisiert auf (-PI, PI]. */
        fun wrapToPi(value: Float): Float {
            var v = value % TWO_PI
            if (v <= -PI) v += TWO_PI
            if (v > PI) v -= TWO_PI
            return v
        }

        /** Normalisiert auf [0, 2*PI). */
        fun wrapTwoPi(value: Float): Float {
            var v = value % TWO_PI
            if (v < 0f) v += TWO_PI
            return v
        }
    }
}

File: ios/Dottie/Sources/Engine/TimingGame.swift
import Foundation

/// 1:1-Port der Kotlin-Engine `TimingGame` (core/.../TimingGame.kt).
///
/// Der Punkt läuft automatisch auf einer Kreisbahn. Irgendwo auf der Bahn
/// liegt eine Zielzone — ein Tap, während der Punkt in der Zone ist,
/// zählt einen Treffer: Die Laufrichtung dreht um, die Zone springt an
/// eine neue Position, das Tempo steigt und die Zone schrumpft. Ein Tap
/// außerhalb der Zone oder ein Überfahren der Zone ohne Tap ist sofort
/// das Ende.
///
/// Alle Konstanten und Formeln stammen unverändert aus dem Original —
/// keine eigene Balance. Winkel in Radiant, Tempo in Radiant pro Sekunde.
final class TimingGame {

    enum Phase {
        case ready
        case running
        case dying
        case over
    }

    /// Reihenfolge entspricht `Twist.entries` in Kotlin — daran hängt die
    /// deterministische Twist-Auswahl im Daily-Modus.
    enum Twist: Int, CaseIterable {
        case pulse
        case drift
        case ghost
        case fake
        case chain
    }

    enum GameEvent: Equatable {
        case started
        case hit
        case perfectHit
        case chainNext
        case twistUnlocked(Twist)
        case died
        case settled
    }

    private var random: KotlinRandom

    private(set) var phase: Phase = .ready

    /// Position des Punkts auf der Bahn.
    private(set) var angle: Float = 0

    /// Laufrichtung: +1 = im Uhrzeigersinn, -1 = dagegen.
    private(set) var direction: Int = 1

    private(set) var zoneCenter: Float = 1.8

    /// Basisbreite der Zone (halb); die effektive Breite kann pulsieren.
    private(set) var zoneHalfWidth: Float = TimingGame.baseZoneHalf

    private(set) var score: Int = 0

    /// Anzahl der Treffer — die Basis für Tempo und Zonenbreite.
    private(set) var hits: Int = 0

    /// Aktuelle Serie perfekter Treffer in Folge.
    private(set) var perfectStreak: Int = 0

    /// Punkte des letzten Treffers, für die Anzeige ("PERFEKT! +3").
    private(set) var lastHitPoints: Int = 0

    private(set) var elapsed: Float = 0

    /// Zeit seit dem letzten Treffer, für Animationen.
    private(set) var timeSinceHit: Float = 99

    /// War der letzte Treffer ein perfekter? Für Anzeige-Effekte.
    private(set) var lastHitPerfect: Bool = false

    /// Die für die aktuelle Zone aktiven Twists.
    private(set) var activeTwists: Set<Twist> = []

    /// Köder-Zone (nur relevant, wenn FAKE aktiv ist).
    private(set) var hasFakeZone: Bool = false
    private(set) var fakeZoneCenter: Float = 0

    /// Wie viele Ketten-Zonen nach der aktuellen noch folgen.
    private(set) var chainRemaining: Int = 0

    /// Drift-Richtung relativ zur Laufrichtung: +1 = flieht, -1 = kommt entgegen.
    private var driftSign: Int = 1

    /// Bereits angekündigte Twists (Banner nur einmal pro Lauf).
    private var announcedTwists: Set<Twist> = []

    /// Events aus tap() werden gepuffert und mit dem nächsten update()
    /// ausgeliefert, damit der UI-Loop eine einzige Event-Quelle hat.
    private var pendingEvents: [GameEvent] = []

    init(random: KotlinRandom = KotlinRandom.systemSeeded()) {
        self.random = random
    }

    /// Relative Position des Punkts zur Zone: negativ = davor, 0 = Mitte.
    func relativeToZone() -> Float {
        return TimingGame.wrapToPi(Float(direction) * (angle - zoneCenter))
    }

    /// Effektive halbe Zonenbreite — pulsiert, wenn PULSE aktiv ist.
    /// sin wie Kotlin/JVM in Double gerechnet und auf Float gerundet,
    /// damit Android und iOS bit-identisch bleiben.
    func effectiveZoneHalf() -> Float {
        if !activeTwists.contains(.pulse) {
            return zoneHalfWidth
        }
        let wave = Float(sin(Double(elapsed * TimingGame.pulseSpeed)))
        let pulse = TimingGame.pulseMinShare + (1 - TimingGame.pulseMinShare) *
            (0.5 + 0.5 * wave)
        return zoneHalfWidth * pulse
    }

    /// Steht der Punkt gerade in der (effektiven) Zielzone?
    var isInZone: Bool {
        return abs(relativeToZone()) <= effectiveZoneHalf()
    }

    /// Ist der Punkt gerade sichtbar? Blinkt nur im GHOST-Twist.
    var isDotVisible: Bool {
        if phase != .running || !activeTwists.contains(.ghost) {
            return true
        }
        let cycle = (elapsed * TimingGame.ghostBlinkSpeed)
            .truncatingRemainder(dividingBy: 1)
        return cycle < TimingGame.ghostVisibleShare
    }

    func currentSpeed() -> Float {
        return min(
            TimingGame.baseSpeed + Float(hits) * TimingGame.speedPerHit,
            TimingGame.maxSpeed
        )
    }

    /// Verarbeitet einen Tap. In READY startet er den Lauf, in RUNNING ist
    /// er der Stopp-Versuch, in OVER (nach kurzer Sperre gegen Wut-Taps)
    /// geht es zurück in den READY-Zustand.
    @discardableResult
    func tap() -> GameEvent? {
        var event: GameEvent?
        switch phase {
        case .ready:
            phase = .running
            elapsed = 0
            spawnZone()
            event = .started
        case .running:
            let rel = relativeToZone()
            let half = effectiveZoneHalf()
            if abs(rel) <= half {
                let perfect = abs(rel) <= half * TimingGame.perfectShare
                registerHit(perfect: perfect)
                event = perfect ? .perfectHit : .hit
            } else if rel > half &&
                rel <= half + currentSpeed() * TimingGame.lateTapForgivenessSeconds {
                // Touch-Latenz-Gnade: Auf der Auslauf-Seite zählt ein
                // minimal verspäteter Tap noch als normaler Treffer.
                registerHit(perfect: false)
                event = .hit
            } else {
                // Auch ein Tap in der Fallen-Zone landet hier: Sie ist
                // mechanisch einfach "daneben" — ihre Gefahr ist optisch.
                die()
                event = .died
            }
        case .dying:
            event = nil
        case .over:
            if elapsed >= TimingGame.restartLockSeconds {
                // Sofort-Neustart: aus der Wut direkt in den nächsten Lauf.
                reset()
                phase = .running
                elapsed = 0
                spawnZone()
                event = .started
            } else {
                event = nil
            }
        }
        if let event = event {
            pendingEvents.append(event)
        }
        return event
    }

    /// Ersetzt die Zufallsquelle — vor jedem Daily-Lauf mit dem Tages-Seed
    /// aufgerufen, damit jeder Versuch des Tages dieselbe Zonen- und
    /// Twist-Abfolge bekommt. `nil` stellt echten Zufall wieder her.
    func reseed(_ seed: Int64?) {
        if let seed = seed {
            random = KotlinRandom(seed: seed)
        } else {
            random = KotlinRandom.systemSeeded()
        }
    }

    /// Setzt alles auf den READY-Zustand zurück (Rekord bleibt beim Store).
    func reset() {
        phase = .ready
        angle = 0
        direction = 1
        zoneCenter = 1.8
        zoneHalfWidth = TimingGame.baseZoneHalf
        score = 0
        hits = 0
        perfectStreak = 0
        lastHitPoints = 0
        elapsed = 0
        timeSinceHit = 99
        lastHitPerfect = false
        activeTwists.removeAll()
        hasFakeZone = false
        chainRemaining = 0
        announcedTwists.removeAll()
        pendingEvents.removeAll()
    }

    /// Schreibt einen Frame fort und liefert die dabei aufgetretenen Events.
    func update(deltaSeconds: Float) -> [GameEvent] {
        let dt = min(max(deltaSeconds, 0), TimingGame.maxDelta)
        elapsed += dt
        timeSinceHit += dt
        var events: [GameEvent] = []
        events.append(contentsOf: pendingEvents)
        pendingEvents.removeAll()

        switch phase {
        case .ready:
            angle = TimingGame.wrapTwoPi(
                angle + Float(direction) * TimingGame.readySpeed * dt
            )
        case .running:
            angle = TimingGame.wrapTwoPi(
                angle + Float(direction) * currentSpeed() * dt
            )
            if activeTwists.contains(.drift) {
                zoneCenter = TimingGame.wrapTwoPi(
                    zoneCenter + Float(direction * driftSign) * TimingGame.driftSpeed * dt
                )
            }
            // Zone ohne Tap überfahren -> vorbei. Geprüft wird gegen die
            // volle Basisbreite, damit eine pulsierende Zone fair bleibt.
            if relativeToZone() > zoneHalfWidth + currentSpeed() * TimingGame.passBufferSeconds {
                die()
                events.append(.died)
            }
        case .dying:
            // Kurzer Freeze für Flash und Shake, dann der Mario-Hüpfer:
            // Das Game-Over-Overlay erscheint erst, wenn der Vogel aus
            // dem Bild gefallen ist (settled = "aufgeschlagen").
            if elapsed >= TimingGame.deathFreezeSeconds + TimingGame.deathFallSeconds {
                phase = .over
                elapsed = 0
                events.append(.settled)
            }
        case .over:
            break
        }
        return events
    }

    private func registerHit(perfect: Bool) {
        hits += 1
        if perfect {
            perfectStreak += 1
            lastHitPoints = min(
                TimingGame.perfectBaseScore - 1 + perfectStreak,
                TimingGame.perfectMaxScore
            )
        } else {
            perfectStreak = 0
            lastHitPoints = 1
        }
        score += lastHitPoints
        timeSinceHit = 0
        lastHitPerfect = perfect
        zoneHalfWidth = max(
            TimingGame.baseZoneHalf - Float(hits) * TimingGame.zoneShrinkPerHit,
            TimingGame.minZoneHalf
        )

        if chainRemaining > 0 {
            // Ketten-Zone: gleiche Richtung, die nächste kommt sofort.
            chainRemaining -= 1
            hasFakeZone = false
            activeTwists.remove(.fake)
            spawnChainZone()
            pendingEvents.append(.chainNext)
        } else {
            direction = -direction
            spawnZone()
        }
    }

    private func spawnZone() {
        // Der Mindestabstand ist zeitbasiert: Egal wie schnell der Punkt
        // schon kreist, bleiben immer mindestens minReactionSeconds bis
        // zur neuen Zone — sonst stirbt man an Physik statt an Skill.
        let minDistance = max(
            TimingGame.minZoneDistance,
            currentSpeed() * TimingGame.minReactionSeconds
        )
        let maxDistance = max(TimingGame.maxZoneDistance, minDistance + 0.4)
        let distance = minDistance + random.nextFloat() * (maxDistance - minDistance)
        zoneCenter = TimingGame.wrapTwoPi(angle + Float(direction) * distance)
        chooseTwists()

        driftSign = random.nextBoolean() ? 1 : -1
        chainRemaining = activeTwists.contains(.chain) ? TimingGame.chainLength : 0

        hasFakeZone = false
        if activeTwists.contains(.fake) {
            let maxFakeDistance = distance - zoneHalfWidth * 3
            if maxFakeDistance > TimingGame.fakeMinDistance {
                let fakeDistance = TimingGame.fakeMinDistance +
                    random.nextFloat() * (maxFakeDistance - TimingGame.fakeMinDistance)
                fakeZoneCenter = TimingGame.wrapTwoPi(
                    angle + Float(direction) * fakeDistance
                )
                hasFakeZone = true
            }
        }
    }

    /// Folge-Zone einer Kette: näher dran, keine neue Twist-Auswahl.
    private func spawnChainZone() {
        let minDistance = max(
            TimingGame.chainMinDistance,
            currentSpeed() * TimingGame.minReactionSeconds
        )
        let maxDistance = max(TimingGame.chainMaxDistance, minDistance + 0.3)
        let distance = minDistance + random.nextFloat() * (maxDistance - minDistance)
        zoneCenter = TimingGame.wrapTwoPi(angle + Float(direction) * distance)
    }

    private func chooseTwists() {
        activeTwists.removeAll()

        // Reihenfolge wie Kotlin `Twist.entries.filter { ... }`.
        let unlocked = Twist.allCases.filter { score >= TimingGame.unlockScore($0) }

        // Ein frisch freigeschalteter Twist wird garantiert gezeigt
        // und einmalig angekündigt.
        if let fresh = unlocked.first(where: { !announcedTwists.contains($0) }) {
            activeTwists.insert(fresh)
            announcedTwists.insert(fresh)
            pendingEvents.append(.twistUnlocked(fresh))
        }

        var shuffled = unlocked
        random.shuffle(&shuffled)
        for twist in shuffled {
            if activeTwists.count >= TimingGame.maxActiveTwists {
                break
            }
            if activeTwists.contains(twist) {
                continue
            }
            if conflictsWithActive(twist) {
                continue
            }
            if random.nextFloat() < TimingGame.twistProbability {
                activeTwists.insert(twist)
            }
        }
    }

    /// Kuratierte Kombis: GEIST + FALLE stapelt fehlende Information mit
    /// tödlicher Fehlinformation — Tode daraus fühlen sich nach Zufall an.
    private func conflictsWithActive(_ candidate: Twist) -> Bool {
        for pair in TimingGame.forbiddenCombos {
            if pair.contains(candidate) &&
                activeTwists.contains(where: { $0 != candidate && pair.contains($0) }) {
                return true
            }
        }
        return false
    }

    private func die() {
        if phase != .running {
            return
        }
        phase = .dying
        elapsed = 0
    }

    // MARK: - Konstanten (identisch zum Kotlin-Original)

    static let maxDelta: Float = 1.0 / 30.0

    // Tempo (Radiant pro Sekunde)
    static let baseSpeed: Float = 2.4
    static let speedPerHit: Float = 0.07
    static let maxSpeed: Float = 5.2
    static let readySpeed: Float = 1.2

    // Zielzone (Radiant)
    static let baseZoneHalf: Float = 0.40
    static let zoneShrinkPerHit: Float = 0.005
    static let minZoneHalf: Float = 0.15
    static let perfectShare: Float = 0.35
    static let minZoneDistance: Float = 1.1
    static let maxZoneDistance: Float = 2.8

    // Fairness (Sekunden)
    static let minReactionSeconds: Float = 0.45
    static let lateTapForgivenessSeconds: Float = 0.07
    static let passBufferSeconds: Float = 0.09

    // Scoring: erster Perfekt +2, jeder weitere in Serie +1 mehr, Deckel +5.
    static let perfectBaseScore: Int = 2
    static let perfectMaxScore: Int = 5

    // Twists
    static let maxActiveTwists: Int = 2

    /// Nie zusammen aktive Twist-Paare.
    static let forbiddenCombos: [Set<Twist>] = [[.ghost, .fake]]
    static let twistProbability: Float = 0.45
    static let pulseSpeed: Float = 5
    static let pulseMinShare: Float = 0.62
    static let driftSpeed: Float = 0.35
    static let ghostBlinkSpeed: Float = 1.6
    static let ghostVisibleShare: Float = 0.62
    static let fakeMinDistance: Float = 0.55
    static let chainLength: Int = 1
    static let chainMinDistance: Float = 1.0
    static let chainMaxDistance: Float = 1.8

    static let deathFreezeSeconds: Float = 0.5

    /// Dauer der Fall-Animation nach dem Freeze (Mario-Hüpfer).
    static let deathFallSeconds: Float = 1.0
    static let restartLockSeconds: Float = 0.55

    /// Ab welchem Score ein Twist ins Spiel kommt.
    static func unlockScore(_ twist: Twist) -> Int {
        switch twist {
        case .pulse: return 5
        case .drift: return 10
        case .ghost: return 15
        case .fake: return 20
        case .chain: return 25
        }
    }

    /// Wichtig: exakt wie Kotlin `(2 * PI).toFloat()` — zum NÄCHSTEN Float
    /// gerundet. Swifts `Float.pi` ist dagegen zur Null hin gerundet und
    /// läge 1 ulp daneben; das würde die Daily-Determinismus-Garantie
    /// gegenüber Android brechen.
    private static let twoPi: Float = Float(2.0 * Double.pi)

    /// Normalisiert auf (-PI, PI]. Die Vergleiche laufen wie in Kotlin
    /// gegen das Double-PI (dort promotet `v <= -PI` den Float).
    static func wrapToPi(_ value: Float) -> Float {
        var v = value.truncatingRemainder(dividingBy: twoPi)
        if Double(v) <= -Double.pi {
            v += twoPi
        }
        if Double(v) > Double.pi {
            v -= twoPi
        }
        return v
    }

    /// Normalisiert auf [0, 2*PI).
    static func wrapTwoPi(_ value: Float) -> Float {
        var v = value.truncatingRemainder(dividingBy: twoPi)
        if v < 0 {
            v += twoPi
        }
        return v
    }
}
